### PR TITLE
Changes to psi4 for MDT interface.

### PR DESCRIPTION
**Pull request by obrien951**
## Description
Changes to psi4 for MDT interface.

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] adds an optional history object to the optimize output. This object contains energies, gradients, and coordinates for the molecule at each step of the optimization. Hooked up to MDT.
  - [x] adds normalmode displacement export to the api and optional saving of normal modes to the wavefunction.
  - [x] Adds handling of several one-electron properties to the API and saves them to the wavefunction.

## Status
- [ ] Ready for comment but not for final merge.

### CODE REVIEW

File: external/upstream/pybind11/CMakeLists.txt [hunk 1]
```diff
@@ -11,6 +11,8 @@ else()
         GIT_TAG v2.0.0
         UPDATE_COMMAND ""
         CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}
+                   -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+                   -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
```
Comment (dgasmith) [inline external/upstream/pybind11/CMakeLists.txt]: This looks like a duplicate of the below?

File: psi4/driver/driver.py [hunk 2]
```diff
@@ -988,6 +988,15 @@ def optimize(name, **kwargs):
         custom_gradient = False
 
     return_wfn = kwargs.pop('return_wfn', False)
+    return_history = kwargs.pop('return_history', False)
+
+
+#Be sure to add wavefunction functionality later once the deep copy issues are worked out
```
Comment (dgasmith) [inline psi4/driver/driver.py]: Comments start with a `# ` and are aligned with other code. There also should not be two spaces above this.

File: psi4/src/psi4/findif/findif.h [hunk 3]
```diff
@@ -47,44 +48,46 @@ namespace findif {
...
 
 // functions to carry out finite-differences
 SharedMatrix fd_1_0(std::shared_ptr<Molecule> mol, Options &options, const py::list& energies);
 //PsiReturnType fd_2_0(std::shared_ptr<Molecule> mol, Options &options, const py::list& energies);
-SharedMatrix fd_freq_0(std::shared_ptr<Molecule> mol, Options &options,
+SharedMatrix fd_freq_0(std::shared_ptr<Molecule> mol, std::shared_ptr<Wavefunction> wfn, Options &options,
                       const py::list& energies, int irrep=-1);
-SharedMatrix fd_freq_1(std::shared_ptr<Molecule> mol, Options &options,
+SharedMatrix fd_freq_1(std::shared_ptr<Molecule> mol, std::shared_ptr<Wavefunction> wfn, Options &options,
                       const py::list& E_list, int irrep=-1);
```
Comment (dgasmith) [inline psi4/src/psi4/findif/findif.h]: We probably do not want to add additional args to this. Instead have the `fd_freq_1` function return a tuple of `(Hessian, freqs)`.

File: psi4/src/psi4/findif/fd_freq_0.cc [hunk 4]
```diff
@@ -258,30 +259,25 @@ SharedMatrix fd_freq_0(std::shared_ptr <Molecule> mol, Options &options,
...
+                vib->lx->set( x, normal_irr[x][i]);
             modes.push_back(vib);
         }
 
-        free(evals);
-        free_block(evects);
-        free_block(normal_irr);
+        //free(evals);
+        //free_block(evects);
+        //free_block(normal_irr);
```
Comment (dgasmith) [inline psi4/src/psi4/findif/fd_freq_0.cc]: Commenting this out will cause memory leaks as these are initialized above.

File: psi4/src/core.cc [hunk 5]
```diff
@@ -139,15 +139,15 @@ PsiReturnType mrcc_load_ccdensities(SharedWavefunction, Options&, const py::dict
...
+                                          int irrep);
 std::vector<SharedMatrix> fd_geoms_freq_1(std::shared_ptr<Molecule>, Options&,
-                                          int irrep = -1);
+                                          int irrep);
 std::vector<SharedMatrix> atomic_displacements(std::shared_ptr<Molecule>, Options&);
 
 SharedMatrix fd_1_0(std::shared_ptr<Molecule>, Options&, const py::list&);
-SharedMatrix fd_freq_0(std::shared_ptr<Molecule>, Options&, const py::list&, int irrep = -1);
-SharedMatrix fd_freq_1(std::shared_ptr<Molecule>, Options&, const py::list&, int irrep = -1);
-SharedMatrix displace_atom(SharedMatrix geom, const int atom,
```
Comment (dgasmith) [inline psi4/src/core.cc]: Why remove the smart defaults here?

File: external/upstream/pybind11/CMakeLists.txt (hunk 1 above)
Comment (loriab) [inline external/upstream/pybind11/CMakeLists.txt]: Yeah, I didn't want Joe to have a non-working codebase if the rebase didn't work. Good catch.

File: psi4/src/psi4/findif/fd_freq_0.cc (hunk 4 above)
Comment (loriab) [inline psi4/src/psi4/findif/fd_freq_0.cc]: We were having real problems with data transport, so I said ok to mem leaks. We'll have to assess lifetime of C-side freq code as to its adherence to good practices.

File: psi4/src/core.cc (hunk 5 above)
Comment (loriab) [inline psi4/src/core.cc]: not sure, should probably go back in as important to irrep choosing.

**Comment by dgasmith** [comment]
These have been moved to #834.

**Pull request closed without merging**